### PR TITLE
#174879639 Build Api for admin to add a country

### PR DESCRIPTION
**Pull request by Fiyin-Anne**
#### What does this PR do

This PR creates a route for admin to add a new country to the database.

#### Description of tasks to be completed?

- verify that the user is an admin
- create a route for admin to add a new country

#### How should this be manually tested/checked?

1. Clone the PR
2. Cd into the branch `ft-admin-can-add-country-174879639`
3. Install the dependencies using `npm install`
4. Set up database connection string as shown in the `.env.sample` file.
5. Run `npm run migrate` to create the database tables
6. Run `npm run seed` to feed the database
7. Start the server by running `npm run start-dev`
8. Sign in via `http://localhost:3000/api/v1/users/signin` using a `POST` method with
```
          {
          "email: "ufuoma@gmail.com",
          "password": 12345
           }
```
and copy the token (without the quotes) from the response body.
9. In a new tab, under the `Authorization` tab, change the `TYPE` to `Bearer Token` and paste the token in the space available. Make a `POST` request to `http://localhost:3000/api/v1/admin/country` to add a country
```
       {
          "nameOfCountry": "togo",
        "gallery":
            "https://image.shutterstock.com/image-illustration/togo-flag-silk-260nw-419363206.jpg",
        "capital": "Lome",
        "population": 205,
        "officialLanguage": "English",
        "region": "West Africa",
        "currency": "CFA franc"
          }

```

#### What are the relevant pivotal tracker stories?
[#174879639](https://www.pivotaltracker.com/story/show/174879639)

#### Architectural Flowchart
[Flowchart: Admin can add country](https://docs.google.com/document/d/1gxqTAKTxTSlh-NUDmC-5dNizilA-YKMvgl2k-vLn-KU/edit?usp=sharing)

### CODE REVIEW

**Review by codeliezel — CHANGES_REQUESTED** [review]
you didn't do what I asked you to on that video call, there should be a function checking whether the user is an admin in the _isAdmin_ middleware.
You have to refactor everything especially for middlewares.
Also, follow instructions.

**Comment by Fiyin-Anne** [comment]
> you didn't do what I asked you to on that video call, there should be a function checking whether the user is an admin in the _isAdmin_ middleware.
> You have to refactor everything especially for middlewares.
> Also, follow instructions.

I created the function in a separate file `/src/services/AdminServices/authServices.js`. Then I called it in `isAdmin`. The functions gets the user's I'd from the token and checks the user's role in the database. You said something about abstraction that's why I separated them.

**Comment by codeliezel** [comment]
> > you didn't do what I asked you to on that video call, there should be a function checking whether the user is an admin in the _isAdmin_ middleware.
> > You have to refactor everything, especially for middlewares.
> > Also, follow instructions.
> 
> I created the function in a separate file `/src/services/AdminServices/authServices.js`. Then I called it in `isAdmin`. The functions get the user's I'd from the token and checks the user's role in the database. You said something about abstraction that's why I separated them.

- abstraction function is different from what I asked you to do.
- I asked you to write a middleware function in the same isAdmin file to check whether a user is an _admin_ and then we use that function in the routes file to precede controllers that have anything to do with an admin.

**Comment by codeliezel** [comment]
Also, why should a user be the one to change the role of a user to an admin before testing. Have a user that is constantly an admin in the users file, then we use that data for sign up or in to check for anything that concerns with an admin.

**Comment by codeliezel** [comment]
npm run seed will migrate all the data, then we already have a pre-filled user that is actually an admin

**Comment by codeliezel** [comment]
refactor your PT story, let there be one scenario that encompasses everything.

**Comment by Fiyin-Anne** [comment]
> > > you didn't do what I asked you to on that video call, there should be a function checking whether the user is an admin in the _isAdmin_ middleware.
> > > You have to refactor everything, especially for middlewares.
> > > Also, follow instructions.
> > 
> > 
> > I created the function in a separate file `/src/services/AdminServices/authServices.js`. Then I called it in `isAdmin`. The functions get the user's I'd from the token and checks the user's role in the database. You said something about abstraction that's why I separated them.
> 
> * abstraction function is different from what and one of the things I asked you to do.
> * I asked you to write a middleware function in the same isAdmin file to check whether a user is an _admin_ and then we use that function in the routes file to precede controllers that have anything to do with an admin.

Okay. I am not sure I understand. The `verifyAdmin` function is what authenticates that a user is an admin, and I already placed it in the `country` route. 

But now, I have also moved the function `checkId`  that checks the user role into the `verifyAdmin `function. The `verifyAdmin` function checks for the token, gets the userId from it, then send it to `checkId` to confirm the role of the user in the database. I called the `checkId` function within the `verifyAdmin`. I have pushed the changes for your review.

**Review by Extremophile-ai — APPROVED** [review]

**Review by bellogo — DISMISSED** [review]

**Review by Bernard-Namangala — CHANGES_REQUESTED** [review]
Fiyin i see in the src/middlewares/index you are importing ./loggedIn but i cant see it in the middlewares maybe you deleted it?

**Review by donaldcrane — APPROVED** [review]

**Review by bellogo — APPROVED** [review]

**Review by Extremophile-ai — APPROVED** [review]

**Review by donaldcrane — APPROVED** [review]

File: src/controllers/country.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        nameOfCountry: countryName, gallery, capital, population, officialLanguage, region, currency
+      };
+      const { error } = validation(newCountry);
+      if (error) return res.status(400).json({ status: 400, error: error.message });
+      const country = await Admin.checkCountry(countryName);
+      if (country) return res.status(409).json({ status: 409, message: "This country already exists in the database." });
+      const createdCountry = await Admin.addCountry(newCountry);
+      return res.status(201).json({ status: 201, message: "A country has been added.", data: createdCountry, });
+    } catch (error) {
+      return res.status(500).json({ status: 500, error });
```
Comment (codeliezel) [inline src/controllers/country.js:25]: This should be `return res.status(500).json({ status: 500, error: "server error or something cool to indicate a deep error" });`

**Review by codeliezel — CHANGES_REQUESTED** [review]
Now, in your eslintrc file, make sure it checks for jsdoc. Now I want you to add jsdoc to every function in our repo.

File: src/controllers/country.js (hunk 1 above)
Comment (Fiyin-Anne) [reply src/controllers/country.js:25]: > Now, in your eslintrc file, make sure it checks for jsdoc. Now I want you to add jsdoc to every function in our repo.

Alright, will work on it now.

**Review by donaldcrane — APPROVED** [review]

**Review by bellogo — APPROVED** [review]

**Review by codeliezel — APPROVED** [review]

**Pull request merged**